site/profile/manifests/core/icinga_resources.pp

```diff
@@ -0,0 +1,832 @@
+  $dns_svc_cmd3   = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${dns_svc_path3}"
+
+  $master_svc_path1 = "${icinga_path}/${master_svc_dhcp_name}.json"
+  $master_svc_cond1 = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${master_svc_dhcp_name}&host=${master_template}' ${lt}"
+  $master_svc_cmd1  = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${master_svc_path1}"
+  $master_svc_path2 = "${icinga_path}/${master_svc_ping_name}.json"
+  $master_svc_cond2 = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${master_svc_ping_name}&host=${master_template}' ${lt}"
+  $master_svc_cmd2  = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${master_svc_path2}"
+  $master_svc_path3 = "${icinga_path}/${master_svc_disk_name}.json"
+  $master_svc_cond3 = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${master_svc_disk_name}&host=${master_template}' ${lt}"
+  $master_svc_cmd3  = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${master_svc_path3}"
+
+  $ipa_svc_path1  = "${icinga_path}/${ipa_svc_name}.json"
+  $ipa_svc_cond1  = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${ipa_svc_name}&host=${ipa_template}' ${lt}"
+  $ipa_svc_cmd1   = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${ipa_svc_path1}"
+  $ipa_svc_path2  = "${icinga_path}/${ipa_svc_ping_name}.json"
+  $ipa_svc_cond2  = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${ipa_svc_ping_name}&host=${ipa_template}' ${lt}"
+  $ipa_svc_cmd2   = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${ipa_svc_path2}"
+  $ipa_svc_path3  = "${icinga_path}/${ipa_svc_disk_name}.json"
+  $ipa_svc_cond3  = "${curl} '${credentials}' -H '${format}' -X GET '${url_svc}?name=${ipa_svc_disk_name}&host=${ipa_template}' ${lt}"
+  $ipa_svc_cmd3   = "${curl} '${credentials}' -H '${format}' -X POST '${url_svc}' -d @${ipa_svc_path3}"
+
+  #Master Host Creation
+  $addhost_path = "${icinga_path}/${master_fqdn}.json"
+  $addhost_cond = "${curl} '${credentials}' -H '${format}' -X GET '${url_host}?name=${master_fqdn}' ${lt}"
+  $addhost_cmd  = "${curl} '${credentials}' -H '${format}' -X POST '${url_host}' -d @${addhost_path}"
+
+  #<---------End Templates Creation----------->
+  #
+  #
+  #<-----Files Creation and deployement------->
+
+  #Create a directory to allocate json files
+  file { $icinga_path:
+    ensure => 'directory',
+  }
+
+  ##Host Templates
+  #Create host template file
+  file { $host_template_path:
+    ensure  => 'present',
+    content => $general_template_content,
+    before  => Exec[$host_template_cmd],
+  }
+  #Add general host template
+  exec { $host_template_cmd:
+    cwd      => $icinga_path,
+    path     => ['/sbin', '/usr/sbin', '/bin'],
+    provider => shell,
+    onlyif   => $host_template_cond,
+    loglevel => debug,
+  }
```

Comment:
Setting aside the fact that I think should shouldn't be in puppet code at all and configured from the director, there is a basic maintainability issue here in that basically, the same two file and exec resources are cut'n'pasted over and over again. This is basically like trying to program without using subroutines/fucntions.  These resources should be encapsulated in a a defined type: https://puppet.com/docs/puppet/6.16/puppet_index.html

Reply:
If we decide to use objects, we would have to create either way one per object. Each exec has it's own condition to trigger or preventing it for triggering